### PR TITLE
[2.x] Filesystem abstraction

### PR DESCRIPTION
**Pull request by NiR-**

### CODE REVIEW

File: src/Knp/Snappy/Filesystem.php [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        }
+    }
+
+    /**
+     * @param string $content
+     * @param string $extension
+     *
+     * @return string
+     */
+    public function createTemporaryFile($content = null, $extension = null): string
```
Comment (polc) [inline src/Knp/Snappy/Filesystem.php]: Maybe we could use http://php.net/manual/fr/function.tmpfile.php instead ?

File: src/Knp/Snappy/Filesystem.php [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    /**
+     * Get the size of the file at $path.
+     *
+     * @param string $path
+     *
+     * @throws FileNotFound
+     *
+     * @return int
```
Comment (Einenlum) [inline src/Knp/Snappy/Filesystem.php]: Maybe add a description about the measure used here

File: src/Knp/Snappy/Filesystem.php [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    public function unlink(string $filename)
+    {
+        if (!$this->exists($filename)) {
+            return;
+        }
+
+        unlink($filename);
+
+        if ($this->exists($filename)) {
+            throw new \RuntimeException(sprintf('Could not delete already existing output file \'%s\'.', $filename));
```
Comment (Einenlum) [inline src/Knp/Snappy/Filesystem.php]: why "output" file?

File: src/Knp/Snappy/Filesystem.php [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    {
+        @mkdir($pathname, 0777, true);
+
+        if (!$this->isDir($pathname)) {
+            throw new \RuntimeException(sprintf('Unable to create directory "%s".', $pathname));
+        }
+    }
+
+    /**
+     * @param string $content
```
Comment (Einenlum) [inline src/Knp/Snappy/Filesystem.php]: string | null (why no typehint btw?)

File: src/Knp/Snappy/Filesystem.php [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+     * @param string $extension
```
Comment (Einenlum) [inline src/Knp/Snappy/Filesystem.php]: string | null

File: src/Knp/Snappy/Filesystem.php (hunk 1 above)
Comment (Einenlum) [inline src/Knp/Snappy/Filesystem.php]: why no typehint?

File: src/Knp/Snappy/Filesystem.php [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+     *
+     * @return string
+     */
+    public function createTemporaryFile($content = null, $extension = null): string
+    {
+        $dir = rtrim($this->temporaryDirectory, DIRECTORY_SEPARATOR);
+
+        if (!is_dir($dir)) {
+            if (false === @mkdir($dir, 0777, true) && !is_dir($dir)) {
```
Comment (Einenlum) [inline src/Knp/Snappy/Filesystem.php]: Why not use directly $this->mkdir?

File: src/Knp/Snappy/Filesystem.php [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+        if (!$this->isDir($pathname)) {
+            throw new \RuntimeException(sprintf('Unable to create directory "%s".', $pathname));
+        }
+    }
+
+    /**
+     * @param string $content
+     * @param string $extension
+     *
+     * @return string
```
Comment (Einenlum) [inline src/Knp/Snappy/Filesystem.php]: `@return string The filename of the temporary file`

File: src/Knp/Snappy/Filesystem.php (hunk 1 above)
Comment (NiR-) [reply src/Knp/Snappy/Filesystem.php]: `tmpfile` create a resource, we don't want that.

**Comment by NiR-** [comment]
@polc Well iirc it does not fully address our needs, thus I'm not keen to add a dependency we would need to extend.

**Pull request merged**